external/llvm-project/mlir/lib/Dialect/LLVMIR/Transforms/SoftwareBf16.cpp

```diff
@@ -0,0 +1,364 @@
+          rawOp, retType, rewriter.getIntegerAttr(retType, newVal));
+      return success();
+    }
+
+    if (auto valDense = val.dyn_cast<DenseElementsAttr>()) {
+      if (valDense.getElementType() != bf16)
+        return failure();
+      DenseElementsAttr newVal = valDense.bitcast(retElemType);
+      rewriter.replaceOpWithNewOp<LLVM::ConstantOp>(rawOp, retType, newVal);
+      return success();
+    }
+
+    if (auto valSparse = val.dyn_cast<SparseElementsAttr>()) {
+      if (valSparse.getElementType() != bf16)
+        return failure();
+      DenseElementsAttr values = valSparse.getValues();
+      DenseElementsAttr newValues = values.bitcast(retElemType);
+      auto newVal = SparseElementsAttr::get(retType.cast<ShapedType>(),
+                                            valSparse.getIndices(), newValues);
+      rewriter.replaceOpWithNewOp<LLVM::ConstantOp>(rawOp, retType, newVal);
+      return success();
+    }
+    // No match otherwise
+    return failure();
+  }
+};
+
+template <typename Op>
+struct BF16AsF32 : OpRewritePattern<Op> {
+  explicit BF16AsF32(MLIRContext *context) : OpRewritePattern<Op>(context) {}
+
+  LogicalResult matchAndRewrite(Op op,
+                                PatternRewriter &rewriter) const override {
+
+    Location loc = op.getLoc();
+    Type opdType = op->getOperand(0).getType();
+    Type opdElementType = opdType;
+    Type i16 = rewriter.getIntegerType(16);
+
+    if (auto opdShaped = opdType.dyn_cast<ShapedType>()) {
+      opdElementType = opdShaped.getElementType();
+    }
+
+    Type resType = op.getResult().getType();
+    Type extType = rewriter.getF32Type();
+    Type resElementType = resType;
+
+    if (auto resShaped = resType.dyn_cast<ShapedType>()) {
+      extType = resShaped.clone(extType);
+      resElementType = resShaped.getElementType();
+    }
+
+    if (resElementType != i16 && opdElementType != i16)
+      return failure();
+
+    llvm::SmallVector<Value, 2> extended;
+    if (isa<LLVM::SIToFPOp>(op) || isa<LLVM::UIToFPOp>(op)) {
+      extended.push_back(op->getOperand(0));
+    } else {
+      for (Value v : op->getOperands()) {
+        extended.push_back(
+            rewriter.create<LLVM::FPExtOp>(loc, extType, v)); // i16->f32
+      }
+    }
+
+    if (resElementType == i16) {
+      Op operation =
+          rewriter.create<Op>(loc, extType, extended, op->getAttrs());
+      rewriter.replaceOpWithNewOp<LLVM::FPTruncOp>(op, resType,
+                                                   operation.getResult());
+    } else { // FCmp
+      rewriter.replaceOpWithNewOp<Op>(op, resType, extended, op->getAttrs());
+    }
+
+    return success();
+  }
+};
+
+struct SoftwareBF16Ext : OpRewritePattern<LLVM::FPExtOp> {
+  explicit SoftwareBF16Ext(MLIRContext *context) : OpRewritePattern(context) {}
+
+  LogicalResult matchAndRewrite(LLVM::FPExtOp op,
+                                PatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
```

Comment:
Can we de-`auto` here? We'll get complained at for good reason (that being, upstream likes `auto` only when not using it would repeat obvious info)